File: wlr/OGUR/OGUR/Creatures/CreatureFactory.cs
using System;
using OGUR.Entities;
using OGUR.Skills;
using SPX.Core;
using SPX.Entities;
using System.Collections.Generic;

namespace OGUR.Creatures
{
    public class CreatureFactory
    {
        private static int __playerCount = 0;
        public static ICreature Create(int actorType, Point2 position)
        {
            AbstractCreature result;
            switch (actorType)
            {
                case OgurActorType.PLAYER:
                    result = new Player(__playerCount++);
                    break;
                default:
                    result = GenerateCreature(actorType);
                    break;
            }
            result.Setup(position);
            EntityManager.AddObject(result);
            return result;
        }

        private static AbstractCreature GenerateCreature(int actorType)
        {
            switch(actorType)
            {
                case OgurActorType.PEON:return new Peon();
                case OgurActorType.ZORB:return new Zorb();
                case OgurActorType.WRATH:return new Wrath();
                case OgurActorType.HAND: return new Hand();
                case OgurActorType.PRIDE: return new Pride();
                case OgurActorType.ENVY: return new Envy();
                case OgurActorType.GLUTTONY: return new Gluttony();
                case OgurActorType.LUST: return new Lust();
                case OgurActorType.SLOTH: return new Sloth();
                case OgurActorType.GREED: return new Greed();
                case OgurActorType.SERPENT: return new Serpent();
                case OgurActorType.BREAKING_WHEEL: return new BreakingWheel();
                default:throw new Exception("No Factory generation logic for: "+actorType);
            }
        }

        public static ICreature CreateRandom(Point2 randomPoint)
        {
            var val = RNG.Rand.Next(0, Generate.Randoms.Count);
            return Create(Generate.Randoms[val], randomPoint);
        }

        public static ICreature CreateMinion(string skillId, ICreature source,SkillEffect effectGraphic=null,Point2 location=null)
        {
            Minion result = null;
            switch (skillId)
            {
                case SkillId.ACID_NOZZLE:result = new AcidNozzle();break;
                case SkillId.DART_TRAP: result = new DartTrap(); break;
                case SkillId.EXPLODE: result = new Explosion(); break;
                case SkillId.ICE_SHARD: result = new IceShard(); break; 
                case SkillId.VAPOR_CLOUD: result = new VaporCloud(); break;
                case SkillId.PLAGUE: result = new PoisonCloud(); break;
                default:
                    throw new Exception("No minion was defined for the given skillId.");
            }
            result.Init(source, effectGraphic);
            if (location != null)
            {
                result.SetLocation(location);
            }
            EntityManager.AddObject(result);
            return result;
        }

        public static void ResetPlayerCount()
        {
            __playerCount = 0;
        }

        public static void IncreasePlayerCount()
        {
            __playerCount ++;
        }

        public static int GetPlayerCount()
        {
            return __playerCount;
        }

        private static List<int> __remainingBosses = new List<int>()
        {
            OgurActorType.WRATH,
            OgurActorType.ENVY,
            OgurActorType.PRIDE,
            OgurActorType.SLOTH,
            OgurActorType.GREED,
            OgurActorType.LUST,
            OgurActorType.GLUTTONY
        };
        public static IEntity CreateNextBoss(Point2 randomPoint)
        {
            int nextBoss = __remainingBosses[RNG.Rand.Next(0,__remainingBosses.Count)];
            __remainingBosses.Remove(nextBoss);
            return Create(nextBoss,randomPoint);
        }
    }
}

File: port-workspace/aigilas-ported/src/com/aigilas/creatures/CreatureFactory.java
package com.aigilas.creatures;import com.xna.wrapper.*;import java.util.*;import com.aigilas.entities.*;import com.aigilas.skills.*;import com.spx.core.*;import com.spx.entities.*;import com.spx.devtools.*;
    public class CreatureFactory
    {
        private static int __playerCount = 0;

        public static ICreature Create(int actorType, Point2 position) throws Exception
        {
            AbstractCreature result;
            switch (actorType)
            {
                case AigilasActorType.PLAYER:
                    result = new Player(__playerCount++);
                    break;
                default:
                    result = GenerateCreature(actorType);
                    break;
            }
            result.Setup(position);
            EntityManager.addObject(result);
            return result;
        }

        private static AbstractCreature GenerateCreature(int actorType) throws Exception
        {
            switch(actorType)
            {
                case AigilasActorType.PEON:return new Peon();
                case AigilasActorType.ZORB:return new Zorb();
                case AigilasActorType.WRATH:return new Wrath();
                case AigilasActorType.HAND: return new Hand();
                case AigilasActorType.PRIDE: return new Pride();
                case AigilasActorType.ENVY: return new Envy();
                case AigilasActorType.GLUTTONY: return new Gluttony();
                case AigilasActorType.LUST: return new Lust();
                case AigilasActorType.SLOTH: return new Sloth();
                case AigilasActorType.GREED: return new Greed();
                case AigilasActorType.SERPENT: return new Serpent();
                case AigilasActorType.BREAKING_WHEEL: return new BreakingWheel();
                default:throw new Exception("No Factory generation logic for: "+actorType);
            }
        }

        public static ICreature CreateRandom(Point2 randomPoint) throws Exception
        {            
            return Create(Generate.Randoms.get(RNG.Next(0, Generate.Randoms.size())), randomPoint);
        }

        public static ICreature CreateMinion(String skillId, ICreature source,SkillEffect effectGraphic,Point2 location) throws Exception
        {
            Minion result = null;
            switch (skillId)
            {
                case SkillId.ACID_NOZZLE:result = new AcidNozzle();break;
                case SkillId.DART_TRAP: result = new DartTrap(); break;
                case SkillId.EXPLODE: result = new Explosion(); break;
                case SkillId.ICE_SHARD: result = new IceShard(); break; 
                case SkillId.VAPOR_CLOUD: result = new VaporCloud(); break;
                case SkillId.PLAGUE: result = new PoisonCloud(); break;
                default:
                    throw new Exception("No minion was defined for the given skillId.");
            }
            result.Init(source, effectGraphic);
            if (location != null)
            {
                result.SetLocation(location);
            }
            EntityManager.addObject(result);
            return result;
        }        public static ICreature CreateMinion(String skillId, ICreature source, SkillEffect effectGraphic) throws Exception        {            return CreateMinion(skillId,source,effectGraphic,null);        }        public static ICreature CreateMinion(String skillId, ICreature source) throws Exception        {            return CreateMinion(skillId,source,null,null);        }

        private static List<Integer> __remainingBosses;        static
        {        	__remainingBosses = Arrays.asList(
            AigilasActorType.WRATH,
            AigilasActorType.ENVY,
            AigilasActorType.PRIDE,
            AigilasActorType.SLOTH,
            AigilasActorType.GREED,
            AigilasActorType.LUST,
            AigilasActorType.GLUTTONY);
        };

        public static int BossesRemaining()
        {
            return __remainingBosses.size();
        }

        public static IEntity CreateNextBoss(Point2 randomPoint) throws Exception
        {
            int nextBoss = __remainingBosses.get(RNG.Next(0,__remainingBosses.size()));
            __remainingBosses.remove(nextBoss);
            return Create(nextBoss,randomPoint);
        }
    }
